spec/services/facility_stats_service_spec.rb

```diff
@@ -0,0 +1,144 @@
+      end
+      uncontrolled.each do |patient|
+        create(:blood_pressure, :hypertensive, patient: patient, facility: patient.assigned_facility, recorded_at: month + 1.month)
+      end
+      [facility1, facility2]
+    end
+  end
+
+  def refresh_views
+    ActiveRecord::Base.transaction do
+      LatestBloodPressuresPerPatientPerMonth.refresh
+      LatestBloodPressuresPerPatientPerQuarter.refresh
+      PatientRegistrationsPerDayPerFacility.refresh
+    end
+  end
+
+  def stats_of_type(stats, type)
+    stats.map { |_, v| v[type] }
+  end
+
+  describe "#call" do
+    it "sets default values for facilities_data and stats_by_size when no facilities are provided" do
+      subject = FacilityStatsService.new(accessible_facilities: [], retain_facilities: [],
+                                         ending_period: period, rate_numerator: "controlled_patients")
+      subject.call
+      expect(subject.facilities_data).to eq({})
+      expect(subject.stats_by_size).to eq({})
+    end
+
+    it "sets data for the past six periods" do
+      facilities = [small_facility1]
+      subject = FacilityStatsService.new(accessible_facilities: facilities, retain_facilities: facilities,
+                                         ending_period: period, rate_numerator: "controlled_patients")
+      subject.call
+      small = subject.stats_by_size["small"]
+      periods = (1..5).inject([period]) { |periods, number|
+        periods << periods.last.previous
+      }
+      expect(small.keys).to match_array(periods)
+    end
+
+    it "sets facilities data for all provided facilities" do
+      facilities = [small_facility1, small_facility2]
+      subject = FacilityStatsService.new(accessible_facilities: facilities, retain_facilities: facilities,
+                                         ending_period: period, rate_numerator: "controlled_patients")
+      subject.call
+      expect(subject.facilities_data.keys).to match_array([small_facility1.name, small_facility2.name])
+      expect(subject.facilities_data.values.map(&:class).uniq).to eq([Reports::Result])
+    end
+
+    it "accurately tallies stats for facilities by size and period" do
+      small_facilities = setup_for_size("small", december - 5.months)
+      medium_facilities = setup_for_size("medium", december - 4.months)
+      large_facilities = setup_for_size("large", december - 3.months)
+      refresh_views
+      all_facilities = small_facilities + medium_facilities + large_facilities
+      subject = FacilityStatsService.new(accessible_facilities: all_facilities, retain_facilities: all_facilities,
+                                         ending_period: period, rate_numerator: "controlled_patients")
+      subject.call
+
+      # all numbers except cumulative_registrations appear in data 3 months after they're recorded
+      expect(stats_of_type(subject.stats_by_size["small"], "controlled_patients")).to eq [0, 0, 0, 2, 0, 0]
```

Comment:
Would `subject.stats_by_size["small"].values` handle this for you? I don't think you need the helper method.

Reply:
Unfortunately, that doesn't quite work because the shape of the data is 
```
{ size: { period: { 'controlled_patients': 0, ...}}}
```
But I've removed the helper anyhow. It wasn't really improving readability.